proc/groupby.go

```diff
@@ -179,81 +179,85 @@ func (g *GroupBy) Pull() (zson.Batch, error) {
 	}
 }
 
-func typeKey(rowkeys []zeek.TypedEncoding) string {
-	var b strings.Builder
-	for _, rowkey := range rowkeys {
-		b.WriteString(rowkey.Type.String())
-		b.WriteByte(';')
-	}
-	return b.String()
-}
-
-func (g *GroupByAggregator) createRow(ts nano.Ts, rowkeys []zeek.TypedEncoding) *GroupByRow {
+func (g *GroupByAggregator) createRow(keyd *zson.Descriptor, ts nano.Ts, vals zval.Encoding) *GroupByRow {
 	// Make a deep copy so the caller can reuse the underlying arrays.
-	v := make([]zeek.TypedEncoding, len(rowkeys))
-	copy(v, rowkeys)
-	key := typeKey(v)
-	_, ok := g.typeCols[key]
-	if !ok {
-		g.typeCols[key] = v
-	}
+	v := make(zval.Encoding, len(vals))
+	copy(v, vals)
 	return &GroupByRow{
-		rowKeys: v,
-		ts:      ts,
-		columns: compile.Row{Defs: g.reducerDefs},
+		keyd:     keyd,
+		keyvals:  v,
+		ts:       ts,
+		reducers: compile.Row{Defs: g.reducerDefs},
 	}
 }
 
-// XXX this could be made more efficient by using exporting zval.Encoding.build
-// and using it rowkey.Body.Build() and do something smarter with the zeek type
-// strings... otherwise we are sending lots of strings to the GC on each record
-// defeating the purpose of g.cacheKey.
-func (g *GroupByAggregator) key(key []byte, rowkeys []zeek.TypedEncoding) ([]byte, error) {
-	if len(rowkeys) > 0 {
-		key = append(key, rowkeys[0].Type.String()...)
-		for _, rowkey := range rowkeys[1:] {
-			key = append(key, ':')
-			key = append(key, rowkey.Type.String()...)
-		}
-		for _, rowkey := range rowkeys {
-			key = append(key, ':')
-			key = append(key, rowkey.String()...)
+func keysTypeRecord(r *zson.Record, keys []GroupByKey) *zeek.TypeRecord {
+	cols := make([]zeek.Column, len(keys))
+	for k, key := range keys {
+		// XXX this needs to recurse the record to find the bottom
+		// column for group-by on record access, e.g., a.b.c should
+		// find the column for "c" by recursing descriptor.Type here
```

Comment:
This comment appears to be out of date?  A fieldexpr resolver does what the comment describes...

Reply:
yes, that was note to self until I figured out how to do it.

Reply:
removed